### PR TITLE
Properties dev

### PR DESCRIPTION
**Pull request by puzhichen**
Add the properties of polarizability, nmr shielding and IR intensity.

### CODE REVIEW

File: gpu4pyscf/properties/tests/test_ir_intensity.py
Comment (wxj6000) [inline gpu4pyscf/properties/tests/test_ir_intensity.py:1]: This is empty.

File: gpu4pyscf/properties/shielding.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            giao = cupy.array(giao)
+            giao_aux = giao[:,:,index]
+            for idirect in range(3):
+                vtmp = np.einsum('pu,p,vp->uv', giao_aux[idirect], wv, ao)
+                vtmp = cupy.ascontiguousarray(vtmp)
+                add_sparse(vmat[idirect], vtmp, index)
+            
+        elif xctype == 'GGA':
+            wv = vxc * weight
+            giao = opt.mol.eval_gto('GTOval_ig', coords.get(), comp=3)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:108]: Is this still running on CPU?

File: gpu4pyscf/properties/tests/test_ir_intensity.py
Comment (puzhichen) [reply gpu4pyscf/properties/tests/test_ir_intensity.py:1]: I will check it.

File: gpu4pyscf/properties/shielding.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+            giao_nabla = opt.mol.eval_gto('GTOval_ipig', coords.get()).reshape(3, 3, -1, nao)
+            giao = cupy.array(giao)
+            giao_nabla = cupy.array(giao_nabla)
+            giao_aux = giao[:,:,index]
+            giao_nabla_aux = giao_nabla[:,:,:,index]
+            for idirect in range(3):
+                # * write like the gpu4pyscf numint part, but explicitly use the einsum
+                aow = cupy.einsum('pn,p->pn', giao_aux[idirect], wv[0])
+                aow += cupy.einsum('xpn,xp->pn', giao_nabla_aux[:, idirect, :, :], wv[1:4])
+                vtmp = cupy.einsum('pn,mp->nm', aow, ao[0])
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:118]: use `contract` in gpu4pyscf.lib, instead of cupy.einsum

File: gpu4pyscf/properties/shielding.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    natom = mf.mol.natm
+
+    shielding_d = cupy.zeros((natom, 3, 3))
+    shielding_p = cupy.zeros((natom, 3, 3))
+    t2 = time.time()
+    h1ao = get_h1ao(mf)
+    t3 = time.time()
+    print("h1ao ", t3-t2)
+    tmp = cupy.einsum('xuv,ua->xav', s1ao, mvir)
+    s1ai = cupy.einsum('xav,vi->xai', tmp, mocc)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:222]: Same time, use contract

File: gpu4pyscf/properties/polarizability.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    nao, nmo = mo_coeff.shape
+    mocc = mo_coeff[:, mo_occ > 0]
+    mvir = mo_coeff[:, mo_occ == 0]
+    nocc = mocc.shape[1]
+    nvir = nmo - nocc
+    vresp = mf.gen_response(mo_coeff, mo_occ, hermi=1)
+
+    def fx(mo1):
+        mo1 = mo1.reshape(-1, nvir, nocc)  # * the saving pattern
+        mo1_mo_real = cupy.einsum('nai,ua->nui', mo1, mvir)
```
Comment (wxj6000) [inline gpu4pyscf/properties/polarizability.py:41]: same here

File: gpu4pyscf/properties/shielding.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+            giao_nabla = cupy.array(giao_nabla)
+            giao_aux = giao[:,:,index]
+            giao_nabla_aux = giao_nabla[:,:,:,index]
+            for idirect in range(3):
+                # * write like the gpu4pyscf numint part, but explicitly use the einsum
+                aow = contract('pn,p->pn', giao_aux[idirect], wv[0])
+                aow += contract('xpn,xp->pn', giao_nabla_aux[:, idirect, :, :], wv[1:4])
+                vtmp = contract('pn,mp->nm', aow, ao[0])
+                vtmp = cupy.ascontiguousarray(vtmp)
+                add_sparse(vmat[idirect], vtmp, index)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:120]: No need to add_sparse twice.

File: gpu4pyscf/properties/tests/test_ir_intensity.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    def test_rks_b3lyp_df(self):
+        print('-------- RKS density fitting B3LYP -------------')
+        e_tot, freq, intensity = run_dft_df_if('B3LYP')
+        assert np.allclose(e_tot, -76.4666819553)
+        qchem_polar = np.array([ [  8.5902540,    -0.0000000,     -0.0000000],
+                                 [  0.0000000,     6.0167648,     -0.0000000],
+                                 [ -0.0000000,    -0.0000000,      7.5688173]])
+        print(freq, intensity)
+        # assert np.allclose(polar, qchem_polar)
```
Comment (wxj6000) [inline gpu4pyscf/properties/tests/test_ir_intensity.py:130]: uncomment assert?

File: gpu4pyscf/properties/shielding.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    # ! imaginary part
+    t1 = time.time()
+    h1ao = -0.5*mf.mol.intor('int1e_giao_irjxp')
+    h1ao += -mf.mol.intor('int1e_igkin')
+    h1ao += -mf.mol.intor('int1e_ignuc')
+    h1ao = cupy.array(h1ao)
+    t2 = time.time()
+    h1ao += get_vxc(mf, dm0)
+    t3 = time.time()
+    print("h1ao 1e: ", t2-t1, "v2e ", t3-t2)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:188]: logging, instead of print

File: gpu4pyscf/properties/shielding.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+    nao = mo_energy.shape[0]
+    idx_occ = mo_occ > 0
+    idx_vir = mo_occ == 0
+    fx = gen_vind(mf, mo_coeff, mo_occ)
+    start_time = time.time()
+    mocc = mo_coeff[:, idx_occ]
+    mvir = mo_coeff[:, idx_vir]
+    s1ao = -mf.mol.intor('int1e_igovlp')
+    t1 = time.time()
+    print("s1ao ", t1-start_time)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:211]: same here

File: gpu4pyscf/properties/shielding.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+    tmp = contract('xuv,ua->xav', s1ao, mvir)
+    s1ai = contract('xav,vi->xai', tmp, mocc)
+    tmp = contract('ned,ue->nud', s1ao, dm0)
+    s1dm = contract('nud,dv->nuv', tmp, dm0)*0.5
+    tmp = contract('xpq,pi->xiq', s1ao, mocc)
+    s1jk = -contract('xiq,qj->xij', tmp, mocc)*0.5
+    tmp = contract('nai,ua->nui', s1jk, mocc)
+    s1jkdm1 = contract('nui,vi->nuv', tmp, mocc.conj())*2
+    t4 = time.time()
+    print('sintegral cal ', t4 - t3)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:230]: Same here

File: gpu4pyscf/properties/shielding.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    print("s1ao ", t1-start_time)
+    dm0 = mf.make_rdm1()
+    natom = mf.mol.natm
+
+    shielding_d = cupy.zeros((natom, 3, 3))
+    shielding_p = cupy.zeros((natom, 3, 3))
+    t2 = time.time()
+    h1ao = get_h1ao(mf)
+    t3 = time.time()
+    print("h1ao ", t3-t2)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:220]: Same here

File: gpu4pyscf/properties/tests/test_ir_intensity.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+
+
+if __name__ == "__main__":
+    print("Full Tests for polarizabillity")
```
Comment (wxj6000) [inline gpu4pyscf/properties/tests/test_ir_intensity.py:135]: Be consistent with file name

File: gpu4pyscf/properties/ir.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    hartree_kj = nist.HARTREE2J*1e3
+    unit2cm = ((hartree_kj * nist.AVOGADRO)**.5 / (nist.BOHR*1e-10)
+               / (2*np.pi*nist.LIGHT_SPEED_SI) * 1e-2)
+    natm = mf.mol.natm
+    nao = mf.mol.nao
+    dm0 = mf.make_rdm1()
+
+    atom_charges = mf.mol.atom_charges()
+    mass = cupy.array([elements.MASSES[atom_charges[i]] for i in range(natm)])
+    intensity = cupy.zeros((3*natm))
```
Comment (wxj6000) [inline gpu4pyscf/properties/ir.py:49]: No need to define `intensity`

File: gpu4pyscf/properties/shielding.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+    omega, alpha, hyb = mf._numint.rsh_and_hybrid_coeff(
+            mf.xc, spin=mf.mol.spin)
+
+    def fx(mo1):
+        mo1 = mo1.reshape(-1, nvir, nocc)  # * the saving pattern
+        mo1_mo_real = contract('nai,ua->nui', mo1, mvir)
+        dm1 = 2*contract('nui,vi->nuv', mo1_mo_real, mocc.conj())
+        dm1 -= dm1.transpose(0, 2, 1)
+        if hasattr(mf,'with_df'):
+            v1 = cupy.zeros((3, nao, nao))
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:50]: `empty` instead of `zeros`

File: gpu4pyscf/properties/shielding.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+            for i in range(3):
+                v1[i] =+mf.get_jk(mf.mol, dm1[i], hermi=2, with_j=False)[1]*0.5*hyb
+        else:
+            v1 = np.zeros((3, nao, nao))
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:54]: Same here

File: gpu4pyscf/properties/shielding.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+            for i in range(3):
+                v1[i] = -jk.get_jk(mf.mol, dm1[i].get(), 'ijkl,jk->il')*0.5*hyb
+            v1 = cupy.array(v1)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:57]: No need to do it here

File: gpu4pyscf/properties/shielding.py [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+        mo_coeff_mask = mo_coeff[index,:]
+        rho = numint.eval_rho2(opt.mol, ao, mo_coeff_mask, mo_occ, None, xctype)
+        vxc = ni.eval_xc_eff(xc_code, rho, deriv=1, xctype=xctype)[1]
+        if xctype == 'LDA':
+            wv = weight * vxc[0]
+            giao = opt.mol.eval_gto('GTOval_ig', coords.get(), comp=3)  # (#C(0 1) g) |AO>
+            giao = cupy.array(giao)
+            giao_aux = giao[:,:,index]
+            for idirect in range(3):
+                vtmp = np.einsum('pu,p,vp->uv', giao_aux[idirect], wv, ao)
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:99]: Those variables are in GPU. Use `contract` instead.

File: gpu4pyscf/properties/shielding.py [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+    idx_vir = mo_occ == 0
+    fx = gen_vind(mf, mo_coeff, mo_occ)
+    mocc = mo_coeff[:, idx_occ]
+    mvir = mo_coeff[:, idx_vir]
+    s1ao = -mf.mol.intor('int1e_igovlp')
+    s1ao = cupy.array(s1ao)
+    dm0 = mf.make_rdm1()
+    natom = mf.mol.natm
+
+    shielding_d = cupy.zeros((natom, 3, 3))
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:206]: No need to define it here.

File: gpu4pyscf/properties/shielding.py [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+    s1dm = contract('nud,dv->nuv', tmp, dm0)*0.5
+    tmp = contract('xpq,pi->xiq', s1ao, mocc)
+    s1jk = -contract('xiq,qj->xij', tmp, mocc)*0.5
+    tmp = contract('nai,ua->nui', s1jk, mocc)
+    s1jkdm1 = contract('nui,vi->nuv', tmp, mocc.conj())*2
+    s1jkdm1 -= s1jkdm1.transpose(0, 2, 1)
+    omega, alpha, hyb = mf._numint.rsh_and_hybrid_coeff(
+            mf.xc, spin=mf.mol.spin)
+    if hasattr(mf,'with_df'):
+        vk2 = cupy.zeros((3, nao, nao))
```
Comment (wxj6000) [inline gpu4pyscf/properties/shielding.py:221]: `empty` instead of `zeros`

**Pull request merged**